File: app/models/decision.py
"""Strict request contracts for the governed Codex decision workflow."""

from __future__ import annotations

import re
from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


PROPOSAL_SCHEMA_VERSION = "codex-proposal-v1"
DEFAULT_PROMPT_VERSION = "codex-decision-v1"


class DecisionAction(str, Enum):
    BUY_NOW = "buy_now"
    CONDITION_ORDER = "condition_order"
    WAIT = "wait"
    AVOID = "avoid"


class DecisionScope(BaseModel):
    model_config = ConfigDict(extra="forbid")

    max_new_positions: int = Field(default=2, ge=0, le=10)
    primary_position_count: int = Field(default=1, ge=0, le=3)

    @model_validator(mode="after")
    def primary_count_fits_scope(self) -> "DecisionScope":
        if self.primary_position_count > self.max_new_positions:
            raise ValueError("primary_position_count cannot exceed max_new_positions")
        return self


class CodexDecisionOverride(BaseModel):
    model_config = ConfigDict(extra="forbid")

    warning_code: str = Field(min_length=1)
    reason: str = Field(min_length=4)
    risk_adjustment: str = Field(min_length=1)


class CodexSelection(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str
    action: DecisionAction
    position_role: Literal["primary", "secondary", "none"]
    requested_quantity: Optional[int] = Field(default=None, ge=0)
    entry_strategy: Optional[Literal["pullback", "breakout", "reference"]] = None
    trigger_price: Optional[Decimal] = Field(default=None, gt=0)
    stop_price: Optional[Decimal] = Field(default=None, gt=0)
    target_price: Optional[Decimal] = Field(default=None, gt=0)
    expires_at: Optional[datetime] = None
    confidence: float = Field(default=0.5, ge=0, le=1)
    thesis: str = Field(min_length=4)
    evidence_refs: list[str] = Field(min_length=1)
    overrides: list[CodexDecisionOverride] = Field(default_factory=list)

    @field_validator("symbol")
    @classmethod
    def normalize_symbol(cls, value: str) -> str:
        digits = re.sub(r"\D", "", str(value or ""))
        if len(digits) != 6:
            raise ValueError("symbol must contain six digits")
        return digits

    @field_validator("evidence_refs")
    @classmethod
    def unique_evidence_refs(cls, values: list[str]) -> list[str]:
        normalized = [str(value).strip() for value in values]
        if any(not value for value in normalized):
            raise ValueError("evidence_refs cannot contain empty values")
        if len(set(normalized)) != len(normalized):
            raise ValueError("evidence_refs must be unique")
        return normalized

    @model_validator(mode="after")
    def validate_action_shape(self) -> "CodexSelection":
        actionable = self.action in {
            DecisionAction.BUY_NOW,
            DecisionAction.CONDITION_ORDER,
        }
        if actionable:
            required = (
                self.requested_quantity,
                self.entry_strategy,
                self.trigger_price,
                self.stop_price,
                self.target_price,
                self.expires_at,
            )
            if any(value is None for value in required) or not self.requested_quantity:
                raise ValueError(
                    "actionable selection requires quantity and price plan"
                )
            if self.position_role == "none":
                raise ValueError("actionable selection requires a position role")
            if not self.stop_price < self.trigger_price < self.target_price:
                raise ValueError(
                    "expected stop_price < trigger_price < target_price"
                )
        else:
            if self.requested_quantity not in (None, 0):
                raise ValueError("wait/avoid selection cannot carry executable quantity")
            if self.position_role != "none":
                raise ValueError("wait/avoid selection position_role must be none")
        return self


class CodexDecisionProposalInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    research_packet_id: str = Field(min_length=1)
    proposal_schema_version: Literal["codex-proposal-v1"] = PROPOSAL_SCHEMA_VERSION
    decision_scope: DecisionScope = Field(default_factory=DecisionScope)
    selections: list[CodexSelection] = Field(default_factory=list)
    portfolio_rationale: str = Field(min_length=4)
    no_action_reason: Optional[str] = None
    prompt_version: str = Field(default=DEFAULT_PROMPT_VERSION, min_length=1)

    @model_validator(mode="after")
    def validate_portfolio_shape(self) -> "CodexDecisionProposalInput":
        if not self.selections and not str(self.no_action_reason or "").strip():
            raise ValueError("no_action_reason is required when selections is empty")

        symbols = [selection.symbol for selection in self.selections]
        if len(set(symbols)) != len(symbols):
            raise ValueError("each symbol may appear only once")

        actionable = [
            selection
            for selection in self.selections
            if selection.action
            in {DecisionAction.BUY_NOW, DecisionAction.CONDITION_ORDER}
        ]
        if len(actionable) > self.decision_scope.max_new_positions:
            raise ValueError("actionable selections exceed max_new_positions")

        primary_count = sum(
            selection.position_role == "primary" for selection in actionable
        )
        if primary_count > self.decision_scope.primary_position_count:
            raise ValueError("primary selections exceed primary_position_count")
        return self


class DecisionConfirmationInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    validation_id: str = Field(min_length=1)
    accepted: bool
    reason: Optional[str] = None

File: frontend/src/api/decision.ts
import { ApiClient } from './request'

export type DecisionAction = 'buy_now' | 'condition_order' | 'wait' | 'avoid'
export type DecisionValidationStatus =
  | 'valid'
  | 'invalid'
  | 'stale_revalidation_required'

export interface DecisionConstraint {
  code: string
  message?: string
  severity?: string
  overrideable?: boolean
  applies_to?: string[]
  source?: string
  symbol?: string
  details?: Record<string, unknown>
}

export interface DecisionEvidence {
  evidence_id: string
  kind: string
  value: unknown
}

export interface DecisionPricePlan {
  entry_strategy?: 'pullback' | 'breakout' | 'reference'
  entry_price?: number | null
  breakout_price?: number | null
  stop_price?: number | null
  target_price?: number | null
  entry_status?: string
  plan_expires_at?: string | null
}

export interface DecisionCandidate {
  symbol: string
  name: string
  identity: {
    code?: string
    name?: string
    objective_segment?: string
    objective_match_score?: number
    [key: string]: unknown
  }
  software_baseline_action: DecisionAction
  software_reason_codes: string[]
  quote: {
    price?: number | null
    source?: string
    trade_at?: string
    quote_checked_at?: string
    status?: string
    actionable?: boolean
  }
  plans: {
    short?: DecisionPricePlan
    swing?: DecisionPricePlan
    position?: DecisionPricePlan
  }
  profile?: Record<string, unknown>
  hard_constraints: DecisionConstraint[]
  soft_warnings: DecisionConstraint[]
  risk_envelope: {
    status?: string
    reason?: string
    lot_size?: number
    max_allowed_quantity?: number
    max_allowed_amount?: number
    max_position_pct?: number
    max_planned_loss_amount?: number
    max_planned_loss_pct_of_assets?: number
  }
  evidence: DecisionEvidence[]
  plan_id?: string | null
}

export interface DecisionResearchPacket {
  research_packet_id: string
  source_baseline_id: string
  as_of?: string
  created_at: string
  material_hash: string
  market_session: {
    phase?: string
    quote_freshness_required_seconds?: number
    [key: string]: unknown
  }
  account: {
    total_assets?: number
    available_cash?: number
    current_exposure_pct?: number
    [key: string]: unknown
  }
  execution_capabilities?: {
    source?: string
    condition_order?: {
      verified?: boolean
      independent_trigger_price_supported?: boolean
      separate_order_limit_price_supported?: boolean
      eligible?: boolean
    }
    market_permissions?: {
      star_market?: {
        verified?: boolean
        tradable?: boolean
        eligible?: boolean
        reason_code?: 'permission_unverified' | 'permission_denied' | null
      }
      beijing_stock_exchange?: {
        verified?: boolean
        tradable?: boolean
        eligible?: boolean
        reason_code?: 'permission_unverified' | 'permission_denied' | null
      }
    }
  }
  market: {
    combined_regime?: string
    domestic_regime?: string
    [key: string]: unknown
  }
  decision_objective: {
    max_new_positions: number
    primary_position_count: number
  }
  hard_risk_policy: {
    available_new_exposure_pct?: number
    hard_single_symbol_cap_pct?: number
    per_position_loss_budget_pct?: number
    total_new_position_loss_budget_pct?: number
    market_red_blocks_new_positions?: boolean
    [key: string]: unknown
  }
  hard_constraints: DecisionConstraint[]
  soft_warnings: DecisionConstraint[]
  candidates: DecisionCandidate[]
  data_quality: {
    unclassified_reason_codes?: string[]
    [key: string]: unknown
  }
  software_baseline: {
    baseline_id: string
    authority: 'software_baseline'
    is_final_decision: false
    market_phase?: string
    revision?: number
    summary?: Record<string, number>
    rule_version?: string
  }
  disclaimer: string
}

export interface SoftwareBaseline {
  decision_id: string
  authority: 'software_baseline'
  is_final_decision: false
  market_phase?: string
  as_of?: string
  summary?: Record<string, number>
  buy_now?: unknown[]
  condition_order?: unknown[]
  wait?: unknown[]
  avoid?: unknown[]
  [key: string]: unknown
}

export interface CodexDecisionOverride {
  warning_code: string
  reason: string
  risk_adjustment: string
}

export interface CodexDecisionSelection {
  symbol: string
  action: DecisionAction
  position_role: 'primary' | 'secondary' | 'none'
  requested_quantity?: number | null
  entry_strategy?: 'pullback' | 'breakout' | 'reference' | null
  trigger_price?: number | string | null
  order_limit_price?: number | string | null
  stop_price?: number | string | null
  target_price?: number | string | null
  expires_at?: string | null
  confidence: number
  thesis: string
  evidence_refs: string[]
  overrides: CodexDecisionOverride[]
}

export interface CodexDecisionProposalPayload {
  research_packet_id: string
  proposal_schema_version: 'codex-proposal-v1'
  decision_scope: {
    max_new_positions: number
    primary_position_count: number
  }
  selections: CodexDecisionSelection[]
  portfolio_rationale: string
  no_action_reason?: string | null
  prompt_version: string
}

export interface CodexDecisionProposal {
  proposal_id: string
  research_packet_id: string
  proposal_schema_version: string
  prompt_version: string
  created_at: string
  status: string
  source: 'codex'
  payload: CodexDecisionProposalPayload
}

export interface DecisionValidationFailure extends DecisionConstraint {
  symbol?: string
}

export interface DecisionValidation {
  validation_id: string
  proposal_id: string
  research_packet_id: string
  validated_at: string
  status: DecisionValidationStatus
  hard_failures: DecisionValidationFailure[]
  accepted_overrides: Array<{
    symbol: string
    warning_code: string
    reason: string
    risk_adjustment: string
  }>
  recalculated: {
    total_cost: number
    total_planned_loss: number
    total_position_weight_pct: number
    selections: Array<{
      symbol: string
      action: DecisionAction
      requested_quantity: number
      total_cost: number
      planned_loss: number
      position_weight_pct: number
    }>
  }
  valid_until?: string | null
  trigger_time_revalidation_required: boolean
  validator_version: string
  disclaimer: string
}

export interface DecisionConfirmation {
  confirmation_id: string
  proposal_id: string
  validation_id: string
  accepted: boolean
  reason?: string | null
  confirmed_at: string
  execution_status: 'not_executed'
  disclaimer: string
}

export interface DecisionWorkspace {
  authority_mode: 'software_baseline' | 'codex_shadow' | 'codex_validated'
  authority: 'software_baseline' | 'codex_validated'
  is_final_decision: boolean
  requires_user_confirmation: boolean
  is_confirmed: boolean
  research_packet: DecisionResearchPacket
  software_baseline: SoftwareBaseline
  codex_proposal: CodexDecisionProposal | null
  validation: DecisionValidation | null
  confirmation: DecisionConfirmation | null
  primary_decision: CodexDecisionProposal | null
  disclaimer: string
}

export interface DecisionConfirmationInput {
  validation_id: string
  accepted: boolean
  reason?: string
}

export const decisionApi = {
  getResearch: (refresh = true) =>
    ApiClient.get<DecisionResearchPacket>('/api/decision/research/today', {
      refresh
    }),

  getBaseline: (refresh = false) =>
    ApiClient.get<SoftwareBaseline>('/api/decision/baseline/today', {
      refresh
    }),

  getFinal: (refresh = false) =>
    ApiClient.get<DecisionWorkspace>('/api/decision/final/today', {
      refresh
    }),

  revalidate: (proposalId: string, refreshQuote = true) =>
    ApiClient.post<DecisionValidation>(
      `/api/decision/proposals/${proposalId}/validate`,
      undefined,
      { params: { refresh_quote: refreshQuote } }
    ),

  confirm: (proposalId: string, payload: DecisionConfirmationInput) =>
    ApiClient.post<DecisionConfirmation>(
      `/api/decision/proposals/${proposalId}/confirm`,
      payload
    )
}
